docs/designs/topology.md

```diff
@@ -0,0 +1,916 @@
+
+- Provide flexible, cluster-agnostic topology hierarchy definition via TopologyDomain CRD
+    - Enable packing constraints for network locality across all Grove scalable resources
+    - Support multiple topology configurations for different environments
+    - Automatic Kueue Topology generation for KAI scheduler integration
+    - Immutable topology configuration ensuring scheduling consistency
+    - Hierarchical constraint validation (child stricter than parent)
+
+## Non-Goals
+
+- Spread constraints across topology domains (ReplicaSpreadDomain)
+    - Root domain constraints for entire resource (RootDomain)
+    - Ratio-based affinity groups between scaling groups (AffinityGroups with PackRatio)
+    - Dynamic topology reconfiguration after creation
+    - Per-workload topology domain selection
+    - Automatic topology inference from workload characteristics
+
+## Proposal
+
+### High-Level Approach
+
+Grove implements topology-aware scheduling through three key components:
+
+1. **TopologyDomain CRD**: Cluster-scoped resource defining topology hierarchy
+    - Admin creates TopologyDomain with ordered list of topology levels
+    - Each level maps friendly name (e.g., "rack") to node label key (e.g., "topology.kubernetes.io/rack")
+    - Multiple TopologyDomains supported for different environments
+
+    2. **Operator Configuration**: References TopologyDomain by name
+        - Operator argument `--topology-domain-name=default` selects which TopologyDomain to use
+        - All workload validation performed against configured TopologyDomain
+        - Enables switching between topologies without changing workloads
+
+    3. **Workload API (TopologyConstraint)**: Users specify packing requirements
+        - PodCliqueSet, PodCliqueScalingGroup, and PodClique each have TopologyConstraint field
+        - Users reference level names from TopologyDomain (e.g., `packDomain: "rack"`)
+        - No direct TopologyDomain reference needed in workloads
+
+### Component Interactions
+
+```
+TopologyDomain CRD ──┐
+  (admin creates)    │
+                     │
+Operator Config ─────┼──> Operator validates PackDomain
+  (--topology-       │    against TopologyDomain.Spec.Levels
+   domain-name)      │
+                     │
+PodCliqueSet ────────┘
+  (packDomain: "rack")
+```
+
+### Automatic Optimization
+
+**Out-of-Box Optimization:**
+
+- Operator automatically generates **preferred** constraints using strictest topology level (e.g., "host")
+    - Applied at all three levels (PodGang, NetworkPackGroup, PodGroup) during translation to scheduler API
+    - Users get optimal packing without configuration
+
+**User Control:**
+
+- Users can specify **required** constraints via `packDomain` for strict placement requirements
+    - Required constraints validated and must be satisfied
+    - Preferred constraints enable best-effort optimization with graceful fallback
+
+### Controller Responsibilities
+
+The TopologyDomain controller manages:
```

Comment:
Why should Grove manage any Kueue CRs? Today KAI scheduler decides to have dependency on Kueue but that is the choice that KAI makes. So while Grove today uses KAI scheduler its responsibility should not be tied to a single scheduler (in this case KAI scheduler). Therefore creation of CRs that are required by KAI should be done in KAI using the enhancements to the `PodGang` scheduler API that Grove defines.

Reply:
It has been clarified and accepted that Grove is not going to generate Kueue Topology CR in the first iteration. When we introduce the scheduler backend framework to support KAI, Volcano and other schedulers then we can take the responsibility to create dependent CRs as required by any specific scheduler.